### ISSUE TITLE
Add support for llamafile model names

### ISSUE THREAD

**Issue by veekaybee** [mozilla-ai/lumigator#228]
Currently llamafiles are supported as generic OAI-compatible servers. While this works, it forces us to specify a valid (i.e. hf-transformers loadable) model name - which is needed for vLLM but might not be the case here as one generally uses GGUF files or llamafile filenames. While not strictly required, these names might be useful to e.g. properly identify the model when reading evaluation results.

Supporting this boils down to adding a llamafile:// prefix to handle llamafiles differently and changing the validation methodology for the provided model names.

**Comment by kernel-loophole** [comment]
hi @veekaybee wanted to work on this .where i can start looking for it .and how the llamafile is gonna identify the model name .

**Comment by aittalam** [comment]
Hi @kernel-loophole, and thanks for your interest!

Our evaluator accepts input configurations like [this one](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/evaluator/examples/configs/evaluation/hf_evaluate_inference_server_config.yaml) for inference servers which are compatible with OAI's API. As you can see, it accepts a `base_url` and an `engine` parameters: the former specifies where to access your llamafile's API, the latter is usually a huggingface-format model name (e.g. `mistralai/Mistral-7B-Instruct-v0.2`). We have not agreed on how to identify a llamafile (e.g. whether it is by the GGUF name or the llamafile's name itself), and AFAIK llamafile server [does not need you to specify any engine](https://github.com/Mozilla-Ocho/llamafile/blob/main/llama.cpp/server/api_like_OAI.py), so I think ATM any string that's meaningful for the user might work.

The main problem ATM is that for this kind of configurations we expect the model to be a valid one, so we need to apply the following changes to support llamafiles: 

1. in evaluator's paths, add support for the llamafile prefix [here](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/evaluator/evaluator/paths.py#L18) and add some validation code for the [AssetPath](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/evaluator/evaluator/paths.py#L41). There is actually no specific rule for validating a llamafile atm so I think a pass would be ok as a first iteration.
2. explicitly add llamafile models in [config_templates](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/backend/backend/config_templates.py) so the name will be mapped to a specific configuration type

Note that (2) is optional, as you can also explicitly provide a configuration template as an input parameter to the API, but I think having something there as an example might be nice: ATM the llamafile:// model you see there does not specify a GGUF or a llamafile filename, but rather the original model - I think that's not good and we should remove it with something that uniquely identifies the correct model.

Thanks again and feel free to reach out if you need more info, I am happy to follow this with you!

**Comment by kernel-loophole** [comment]
thanks @aittalam for the detailed explanation!

I see the points you're raising regarding the current configuration setup. Supporting llamafiles by adding a prefix in the evaluator's paths makes sense as a first step, and I agree that for now, having a pass for the validation of the AssetPath seems like a good approach. We can iterate on this later as needed.

Regarding the explicit configuration templates, I think it’s a great idea to include an example to help clarify the setup. Even if it’s optional, having something mapped to a specific configuration type will make it easier to avoid confusion in the future. I also agree with removing the current llamafile:// placeholder in favor of something that directly identifies the right model, particularly as we figure out a convention for naming llamafiles (GGUF name or llamafile name itself).

I’ll start working on these changes, and feel free to reach out if you have any further thoughts or suggestions!

**Comment by aittalam** [comment]
> I’ll start working on these changes, and reach out if i have any further thoughts or suggestions!

That's great, many thanks!

**Comment by aittalam** [comment]
Hi @kernel-loophole, I think it might be related to Docker authentication. I found the following two links which might help:

- https://www.shkodenko.com/how-to-fix-pass-store-is-uninitialized-on-ubuntu-linux-in-docker-setup/
- https://stackoverflow.com/questions/50151833/cannot-login-to-docker-account

Let me know whether any of the suggested solutions work for you!

**Comment by kernel-loophole** [comment]
@aittalam thanks i config the pass store and run it .but this is taking too much time to build the image .40 min and still building 
![Screenshot from 2024-10-14 06-52-14](https://github.com/user-attachments/assets/1a0f7d63-51c5-40eb-953b-689400cc29b7)

**Comment by kernel-loophole** [comment]
@aittalam need guideline on this ,when i run this without sudo `permission denied while trying to connect to the Docker daemon socket ` and when run with sudo then is taking too much time 
![Screenshot from 2024-10-15 01-00-27](https://github.com/user-attachments/assets/2c76d1d8-69e1-4732-8614-d9064e58e4cf)

**Comment by aittalam** [comment]
@kernel-loophole thanks for catching this, if you are using linux and your distribution installs `docker` so that it can only be run by root you definitely need sudo. We should add this to the documentation!

Regarding the amount of time required for building, the first time you run `make local-up` it will have to download docker images for the different components and then install python packages, so one can expect to download a few GBs. Do you have a way to track your download speed? I use `bpytop` for that so I can get a sense of whether the download is going well or if it is hanging somewhere.
Regardless of this, we should definitely warn about this large download in the documentation so people know what to expect in this phase.

**Comment by kernel-loophole** [comment]
@aittalam thanks ,yeah sure .will start the downloading again and see what i can get .

**Comment by kernel-loophole** [comment]
hi @aittalam run this for long time but still go nothing 
![Screenshot from 2024-10-16 07-30-38](https://github.com/user-attachments/assets/18b5c118-4f05-42fe-82ff-289fa51b0dd1)

**Comment by aittalam** [comment]
Hi @kernel-loophole! I have tested the latest code both on Mac and on linux (which is your use case, is that correct?).
This is what I get:

Mac: 
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/bfeed614-0b06-4bed-a955-1975de7e0339">

Linux:
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/8c338f6c-8d4c-4bbb-b7a5-4ef19b98dc4a">

I now realise the issue is not at the download stage, otherwise you would not have gotten there - that step occurs after the containers are pulled. I will look into that as in my case that "load metadata" step is almost immediate.

As a safety check, can you please try to manually connect to the two addresses that the process seems to be stuck on? In theory you should be able to connect to them via browser. I will look for similar issues in the meantime - thanks again for sharing these issues!

**Comment by aittalam** [comment]
@kernel-loophole one more request: could you please share with me the specs of the machine you are running the build on? I'll try to replicate on a VM with similar specs to see if there's anything related to the HW configuration 🙏

**Comment by kernel-loophole** [comment]
@aittalam sure will do and i will try to run this

**Comment by kernel-loophole** [comment]
here are the spec for this 
![Screenshot from 2024-10-17 06-21-49](https://github.com/user-attachments/assets/ee23c404-9502-4493-86de-43924ee00950)

**Comment by aittalam** [comment]
I think your resources should be enough! I have seen there are a few different causes for the specific issue you get, sometimes it's about network configuration/ [domain name resolution](https://github.com/docker/for-win/issues/10247) and sometimes they fixed errors (not delays though) in the same phase by [reauthenticating](https://forums.docker.com/t/help-on-error-internal-load-metadata-for-docker-io-library-node-18-alpine/144360) (that's something you have already done though). Still trying to replicate this, but we'll find a solution for it.

**Comment by kernel-loophole** [comment]
@aittalam finally able to run it .there was configuration problem with the `pass`.will start adding support soon  
![Screenshot from 2024-10-18 00-44-14](https://github.com/user-attachments/assets/0efff803-de68-4960-b60b-2ad75a2382fd)

**Comment by kernel-loophole** [comment]
thanks for your help @aittalam

**Comment by kernel-loophole** [comment]
@aittalam was trying to run this notebook https://github.com/mozilla-ai/lumigator/blob/main/notebooks/walkthrough.ipynb .any idea about this .there is no url defined for this route 
![Screenshot from 2024-10-21 08-17-38](https://github.com/user-attachments/assets/3a29d93c-97ba-423b-9089-fc77226ff209)

**Comment by kernel-loophole** [comment]
okay sure .will create a issue for this

**Comment by kernel-loophole** [comment]
@aittalam make here are the changes i have make .can you review https://github.com/mozilla-ai/lumigator/compare/main...kernel-loophole:lumigator:main

**Comment by javiermtorres** [comment]
Hi @kernel-loophole , we're preparing an SDK to use the lumigator endpoints more easily. Please check these steps (can be done in a jupyter notebook):

* First import the SDK dependencies (this will not be necessary in the final SDK, but currently we're publishing only to the test pypi repo):
  * `pip install loguru requests pydantic`
* Import the SDK package from testpypi:
  * `pip install -i https://test.pypi.org/simple/ lumigator-sdk==0.1.0`
* Get the following sample dataset (just 5 rows) that @aittalam mentioned (there's a longer dataset, but I think we still need to publish it as part of the tests):
  * https://raw.githubusercontent.com/mozilla-ai/lumigator/refs/heads/main/lumigator/python/mzai/sample_data/dialogsum_exc.csv
* In the repository, start the containers using `start-lumigator-build` and wait for all the containers to be ready.
* Run the following script:
```
from pathlib import Path

from lumigator_schemas.datasets import DatasetFormat
from lumigator_sdk.lumigator import LumigatorClient

lumi_client = LumigatorClient("localhost:8000")
dialog_data = "dialogsum_exc.csv"

with Path.open(dialog_data) as file:
    datasets = lumi_client.datasets.get_datasets()
    print(datasets)
    # assert datasets.total == 0
    dataset = lumi_client.datasets.create_dataset(dataset=file, format=DatasetFormat.EXPERIMENT)
    print(dataset)
    datasets = lumi_client.datasets.get_datasets()
    # assert datasets.total == 1
    print(datasets)
    dataset = lumi_client.datasets.get_dataset(datasets.items[0].id)
    assert dataset is not None
    lumi_client.datasets.delete_dataset(datasets.items[0].id)
    datasets = lumi_client.datasets.get_datasets()
    assert datasets.total == 0
```

This should upload the dataset. We're currently changing the job endpoint, but we'll shortly finish the job endpoint refactoring and we'll be able to complete this short example with the evaluation that @aittalam mentioned. This may be included as an example in the docs, if people find it useful.

**Comment by kernel-loophole** [comment]
@javiermtorres thanks for providing the detail information .I'll go through the steps and check out the SDK in a Jupyter notebook.. If I encounter any issues I'll let you know.

**Comment by kernel-loophole** [comment]
@javiermtorres while trying ` pip install -i https://test.pypi.org/simple/ lumigator-sdk==0.1.0`  ,the package is missing `ERROR: Could not find a version that satisfies the requirement lumigator-sdk==0.1.0 (from versions: none)
`

**Comment by javiermtorres** [comment]
Funny... Could you please try installing directly using the wheel at https://test.pypi.org/project/lumigator-sdk/#files ?

**Comment by kernel-loophole** [comment]
hi @javiermtorres i find that the problem was the with python version and update the python version to `3.11.0` and it work .but when i try to run this script cause `connection refused` .also when i run the `make start-lumigator` it start all the containers .when run through the ` make local-up` ,cause `ERROR [backend stage-0 5/5] RUN uv sync --frozen `.before that the localhost was working.

**Comment by javiermtorres** [comment]
Can you try running `make start-lumigator-build`instead, just in case?

**Comment by kernel-loophole** [comment]
okay sure

**Comment by kernel-loophole** [comment]
hi @javiermtorres run using the above command and work .run the script you provided but return empty .
![Screenshot from 2024-10-28 07-55-33](https://github.com/user-attachments/assets/442f3193-e11f-47d9-9aca-0a6ad67daa8f)

2024-10-28 07:51:23.299 | ERROR    | lumigator_sdk.client:_make_request:56 - Request failed: 422 Client Error: Unprocessable Entity for url: http://localhost:8000/api/v1/datasets/

requests.exceptions.HTTPError: 422 Client Error: Unprocessable Entity for url: http://localhost:8000/api/v1/datasets/

**Comment by javiermtorres** [comment]
Does it run ok then? The datasets should then be available inside lumigator.

**Comment by kernel-loophole** [comment]
yes it work fine as described

**Comment by kernel-loophole** [comment]
this issues has been solved #321
